Develop tk

Skyblueballykid opened the pull request:
I added a Dockerfile (I tried to follow Skylar's instructions) and updated docker-compose, let me know what you guys think.

Skyblueballykid committed:
updates to docker-compose file

Skyblueballykid committed:
add Dockerfile

skyripley commented on docker-compose.yml:
```diff
@@ -1,6 +1,16 @@
 
 services:
+  web:
+    build: ./backend
```
```suggestion
    build: 
      context: ./backend
```

Skyblueballykid replied:
Done in latest commit.

skyripley commented on Dockerfile:
```diff
@@ -0,0 +1,7 @@
+FROM python:3
```
We need to move this file to the `backend` directory

Skyblueballykid replied:
Done in latest commit.

skyripley commented on docker-compose.yml:
```diff
@@ -1,6 +1,16 @@
 version: "3"
 
 services:
+  web:
```
```suggestion
  backend:
    container_name: backend
```

Skyblueballykid replied:
Done in latest commit.

skyripley commented on docker-compose.yml:
```diff
@@ -1,6 +1,16 @@
+    build: ./backend
+    command: python manage.py runserver 0.0.0.0:8000
+    volumes:
+    - .:/var/app
```
```suggestion
    - ./backend:/var/app
```

Skyblueballykid replied:
Done in latest commit.

skyripley commented on docker-compose.yml:
```diff
@@ -1,6 +1,16 @@
+    ports:
+        - "8000:8000"
+    depends_on:
+        - postgres
```
I noticed that there's a race condition between postgres actually starting and django trying to connect to it. This `depends_on` will make the `web` (or `backend`) service wait for the `postgres` container to start, but it doesn't wait for postgres itself to start. This means that occasionally when we run `docker-compose up --build`, or whatever command you're using, django will fail to connect to the db. One workaround would be to start the services separately and make sure that postgres itself starts before starting the backend service. Seems like some people recommend using https://github.com/vishnubob/wait-for-it or adding a health check to the `postgres` service and forcing the `backend` service to wait for it to be healthy, but I'm curios if @ntt2k might already know of a good method since he's familiar with django

Skyblueballykid replied:
Still open.

skyripley commented on docker-compose.yml:
```diff
@@ -1,6 +1,16 @@
+    volumes:
+    - .:/var/app
+    ports:
+        - "8000:8000"
```
```suggestion
    ports:
        - "8000:8000"
    environment:
      - POSTGRES_HOST=postgres
```

skyripley replied:
This will make it so that we try to connect to the `postgres` container rather than localhost. Connecting to localhost will fail since we have two separate containers. 

Another option would be to create a `.env.example` file in the base directory of the project that we copy to `.env` before running docker-compose, add `POSTGRES_HOST=postgres` to that file, and add `env_file: ./.env` here so that we pull our "secrets" from an ephemeral file. Then we'd need to add `.env` to our `.gitignore` so that no one accidentally commits production secrets, and also move all of our other secrets here to that file. I'm fine with just doing the suggestion above though for now and we can make our secrets management better in another PR if we think it's necessary since this is probably a nice-to-have for a grad school project.

Skyblueballykid replied:
Done in latest commit.

skyripley reviewed (changes requested):
Can you also add some info to the readme so that everyone knows how to run this locally with `docker-compose`? And let me know when you're finished with the changes so that I can re-review (or you can re-request a review)

Skyblueballykid commented:
Cool, thanks for all the feedback. I will review when I get a chance and make the suggested changes...I might have some follow up questions.

Skyblueballykid committed:
suggested changes to dockerfile and docker-compose

Skyblueballykid commented:
I updated the docker-compose file to reflect Skylar's comments and moved the Dockerfile to backend. Let me know what you guys think, the YAML validates, but hopefully I got the indentation of everything right.

skyripley committed:
Fix indentation and volume
